[BEAM-7450] Support unbounded reads with HCatalogIO

jhalaria opened the pull request:
This PR adds the ability to add unbounded amounts of data using HCatalogIO

1.  It extends PollFn which is responsible for emitting new partition indices when they become available
2. A reader which will read the partition that was jus emitted.

R: @iemejia @lukecwik 
CC: @jkff  (just for info or glad to receive feedback if you are interested).

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

jhalaria commented:
@iemejia - Please review when you get a chance. Thank you.

apilloud commented:
Test failure looks real.
```

java.lang.IllegalArgumentException: org.apache.beam.sdk.io.hcatalog.PartitionPoller, @ProcessElement processElement(ProcessContext, PartitionRangeTracker): Has tracker type PartitionRangeTracker, but the DoFn's tracker type must be of type RestrictionTracker.
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures$ErrorReporter.throwIllegalArgument(DoFnSignatures.java:1495)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures$ErrorReporter.checkArgument(DoFnSignatures.java:1500)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.verifySplittableMethods(DoFnSignatures.java:595)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.parseSignature(DoFnSignatures.java:474)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.lambda$getSignature$0(DoFnSignatures.java:140)
	at java.util.HashMap.computeIfAbsent(HashMap.java:1127)
	at org.apache.beam.sdk.transforms.reflect.DoFnSignatures.getSignature(DoFnSignatures.java:140)
	at org.apache.beam.sdk.transforms.ParDo.validate(ParDo.java:547)
	at org.apache.beam.sdk.transforms.ParDo.of(ParDo.java:397)
	at org.apache.beam.sdk.io.hcatalog.HCatalogUnboundedReader.expand(HCatalogUnboundedReader.java:110)
	at org.apache.beam.sdk.io.hcatalog.HCatalogUnboundedReader.expand(HCatalogUnboundedReader.java:77)
	at org.apache.beam.sdk.Pipeline.applyInternal(Pipeline.java:537)
	at org.apache.beam.sdk.Pipeline.applyTransform(Pipeline.java:471)
	at org.apache.beam.sdk.values.PCollection.apply(PCollection.java:357)
	at org.apache.beam.sdk.io.hcatalog.HCatalogIOTest.testWriteThenUnboundedReadSuccess(HCatalogIOTest.java:220)
```

iemejia commented:
I still need to check in more details the Readers and the intention. Not sure if I understood well but I have the impression that the Bounded could be eventually a replacement of HCatalogIO current Read. Is this the case? If so probably it would be easier to divide this PR in two (one that introduces the Bounded SDF version) and another for the Unbounded but well maybe I have to do a second round to get this correctly.

jhalaria commented:
@iemejia - Thank you so much for your feedback.

- I have addressed all of your recommendations from above.
- Removed the extra tracker and used offsetRangeTracker instead.
- Unit test coverage for the PartitionRangeTracker.
- Refined the public API to be HcatalogIO.unbounded()
- This PR only adds an unbounded reads capability. In order for the unit tests to not continuously poll for new partitions, I had to add a flag to make the unbounded behave as bounded.
- I will have a subsequent PR that unifies the two readers into one and the caller can decide which one to use based a flag. If we like this approach, then we can potentially have just one `reads` API. However the only down side of doing this would be adding certain checks that allows reading a table even without partitions which is a requirement for the unbounded reader to work correctly.


Please have another look at the PR and let me know your feedback. Thanks.

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatRecordReaderFn.java:
```diff
@@ -0,0 +1,204 @@
+
+  private long getFileSizeForPartitions(PartitionWrapper partitionWrapper) throws Exception {
+    IMetaStoreClient client = null;
+    Configuration conf = new Configuration();
```
Can you extract the code that creates a new HiveMetastoreClient into a new static method in a class called `HCatalogUtils` and use it here and in `PartitionPollerFn` too. Extra points if you change it in `HCatalogIO.getEstimatedSizeBytes` too.

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -123,8 +132,132 @@ public static Read read() {
+      checkArgument(
+          readSpec.getPartitionComparator() != null, "withPartitionComparator() is required");
+      return input
+          .apply("ConvertToReadRequest", Create.of(new ArrayList<>(Arrays.asList(readSpec))))
```
`.apply("ConvertToReadRequest", Create.of(readSpec))` simpler = better.

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java:
```diff
@@ -0,0 +1,146 @@
+    for (int i = trueStart; i < forSort.size(); i++) {
+      if (tracker.tryClaim(forSort.get(i))) {
+        c.output(
+            new HCatRecordReaderFn.PartitionWrapper(
```
I am not sure that we need the `PartitionWrapper` class. Can't we just add the missing part of the `Read`, the partition as an attribute of the readRequest (with package private access) and then just pass the readRequest (`Read`) object all around?

jhalaria replied:
Good point. After the refactor, since everything is now in Read(s), this can go away.

jhalaria commented:
@iemejia - I think once this PR gets merged, the next step will clean up the multiple reads method that we currently have. Moving forward there will only be one Read class that can read both bounded and unbounded data depending on the parameters passed. Do you think we can tackle that as the next step or if not what would make it cleaner in this PR? 
When I am extending from Read, I have to add the boilerplate code for the properties inside Read. I thing that we can do in this PR itself is to have only 1 Reads class and then within the expand, call the DoFn's needed for unbounded reads if only the right arguments are set. What do you think?

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatRecordReaderFn.java:
```diff
@@ -0,0 +1,165 @@
+    return new OffsetRange(0, readerContext.numSplits());
+  }
+
+  @NewTracker
```
This is unnecessary to declare, the default implementation will do the same thing.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatRecordReaderFn.java:
```diff
@@ -0,0 +1,165 @@
+    }
+    ReaderContext readerContext = getReaderContext(processContext.element(), desiredSplitCount);
+
+    for (long i = tracker.currentRestriction().getFrom(); tracker.tryClaim(i); i++) {
```
Hmm why does this one need to be an SDF rather than simply a composition of two DoFn's with a reshuffle in between? (where the first DoFn would emit (Read, int) pairs where the int is 0..numSplits) Is there so many splits that it is prohibitive?

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -123,12 +129,147 @@ public static Read read() {
+   * Read unbounded data from Hive. This essentially means that as new partitions show up, we start
+   * reading data for that partition.
+   */
+  public static UnboundedRead unbounded() {
```
Are you sure you need a new class for this, can you just add a boolean parameter to Read "isUnbounded" and add an if() statement in expand?

iemejia replied:
That was probably an ill advise of my part (there was  a totally different approach, not the read based one too). @jhalaria I think this is probablysimilar in principle to what we called the unification, maybe worth to tackle it for this PR.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java:
```diff
@@ -0,0 +1,141 @@
+
+/** Unbounded poller to listen for new partitions. */
+@UnboundedPerElement
+class PartitionPollerFn extends DoFn<Read, Read> {
```
Can this one be replaced with a Watch transform, whose PollFn returns a complete list of current partitions? it scales quite well (if I remember correctly, e.g. a million elements was not a problem, as long as you encode them efficiently), and seems like it would save you a huge amount of code in this PR.

iemejia replied:
Just curiosity question @jkff, does this mean then that we won't need an Unbounded SDF then? How do you tackle the case when new partitions appear?
(KafkaIO has an open ticket on detecting new partitions so I was wondering if the approach you propose could work there too).

jkff replied:
Yes, it does mean that we don't need an SDF here at all; and yes, this would likely work for KafkaIO too. Watch would poll for the current list of partitions, and emit new ones as they appear without re-emitting the already known ones, just like when we watch for new files.

It also means that the current PR would not be introducing any new SDFs, but instead merely reusing an existing one (Watch) - which is a very good thing and in line with SDF's position as enabling more composable/reusable IO APIs as opposed to people having to reimplement similar features in different IOs.

iemejia replied:
Wow, that's super awesome, thanks for bringing this idea @jkff !

iemejia replied:
@jhalaria For further clarity what @jkff is suggesting is to use `Watch` to follow a pattern like [`FileIO.Match.continously`](https://github.com/apache/beam/blob/2798d626507c7f5c8699096b3a188200ac64a299/sdks/java/core/src/main/java/org/apache/beam/sdk/io/FileIO.java#L627) does to poll and discover the new partitions.

jhalaria commented:
Hey @jkff , Thank you for your suggestion. It would definitely make this a lot simpler. I just wanted to make sure that I understand your overall solution correctly:

- Add a simple Watch transform that outputs all partitions List<Partitions> whenever it gets invoked. Signature of Transform 
```java 
PartitionWatcher extends PollFn<HCatalogIO.Read, List<Partition>> ```
- A DoFn that will invoke HcatalogIO.read with a filter condition added that can read the partitions.

jkff commented:
@jhalaria Yes, I'm suggesting that you replace ParDo.of(new PartitionPollFn) with Watch.growthOf(new PartitionPollFn), and change PartitionPollFn to be a Watch.PollFn instead of a DoFn.

(Note that this will also make your tests much simpler: now you only need to verify that your PollFn correctly returns the results of a single poll - the rest is already tested in Watch)

jhalaria commented:
@jkff  @iemejia - This is much cleaner after implementing @jkff suggestions. Should be ready for review now.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -186,6 +216,26 @@ public Read withFilter(String filter) {
       return toBuilder().setFilter(filter).build();
     }
 
+    /** Sets the duration after which the PartitionPollerFn resumes polling for new partitions. */
```
This documentation ideally should be phrased in terms that make sense to a client, rather than reference implementation details. E.g. "If specified, poll for new partitions at this periodicity. Then the returned PCollection will be unbounded. If not specified, (what happens then?)"

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -186,6 +216,26 @@ public Read withFilter(String filter) {
+      return toBuilder().setPollingInterval(pollingInterval).build();
+    }
+
+    /** Specifies the name of the column which contains the watermark values. */
```
This comment looks wrong. I'm generally not sure what this property is for.

jhalaria replied:
Removed. Older property. Not needed.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -196,11 +246,31 @@ Read withContext(ReaderContext context) {
+        checkArgument(getPollingInterval() != null, "withPollingInterval() is required");
+        final Read readSpec =
+            new AutoValue_HCatalogIO_Read.Builder()
+                .setDatabase(DEFAULT_DATABASE)
```
Should this really always use DEFAULT_DATABASE?

jhalaria replied:
I didn't really need to do this here. Removed this logic of creating a spec from here. However, the current version of HcatalogIO does use `DEFAULT_DATABASE`. I can change it to not do that if we want.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -196,11 +246,31 @@ Read withContext(ReaderContext context) {
+            .apply(Reshuffle.of())
+            .apply("PartitionReader", ParDo.of(new PartitionReaderFn()));
+      } else {
+        // Treat as Bounded
```
If the only difference is that either we poll for new partitions or we don't, then we don't need shouldTreatUnboundedAsBounded, and should simply check whether pollingInterval is specified.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java:
```diff
@@ -0,0 +1,71 @@
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.joda.time.Instant;
+
+/** Unbounded poller to listen for new partitions. */
```
Boundedness or unboundedness do not apply to a poll function, it simply returns the current state of the thing being polled.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java:
```diff
@@ -0,0 +1,71 @@
+        .collect(Collectors.toList());
+  }
+
+  @Setup
```
Poll fns don't have setup or teardown, you need to put this into apply() itself.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionSplitterFn.java:
```diff
@@ -0,0 +1,36 @@
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+
+/** Unbounded reader for doing reads from Hcat. */
```
This comment looks wrong.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionSplitterFn.java:
```diff
@@ -0,0 +1,36 @@
+
+  /** Splits partition into key value pairs. */
+  @ProcessElement
+  @SuppressWarnings("unused")
```
What's unused here?

jhalaria replied:
Intellij complains about processElement as unused. It was more to make intellij happy. Will remove it.

jkff commented on sdks/java/io/hcatalog/src/test/java/org/apache/beam/sdk/io/hcatalog/HCatalogIOTest.java:
```diff
@@ -163,6 +179,55 @@ public void processElement(ProcessContext c) {
+                    .withDatabase(TEST_DATABASE)
+                    .withPartitionCols(partitions)
+                    .withTable(TEST_TABLE)
+                    .withPollingInterval(Duration.millis(15000)))
```
Does this test terminate? I think your code sets up Watch.growthOf() in a way that doesn't terminate..

jhalaria replied:
It didn't earlier. Added the ability to specify a termination condition

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogUtils.java:
```diff
@@ -0,0 +1,79 @@
+    int desiredSplitCount = 1;
+    long estimatedSizeBytes = HCatalogUtils.getFileSizeForPartitions(readRequest);
+    if (estimatedSizeBytes > 0) {
+      desiredSplitCount = (int) Math.ceil((double) estimatedSizeBytes / Integer.MAX_VALUE);
```
What's the motivation for this logic?

jhalaria replied:
While reading a partition, we calculate the size of the partition, so that we can set the value for HCatConstants.HCAT_DESIRED_PARTITION_NUM_SPLITS

jkff replied:
Sorry, I mean, why is the correct value for desiredSplitCount = estimatedSizeBytes/Integer.MAX_VALUE? Is it cause we want every split to be of size Integer.MAX_VALUE? If so, why? That seems pretty large.

jhalaria replied:
Ah. I see. I fixed it to have the desired bundle size has 128 MB as default. This is only a hint to the DataTransfer API(s). The actual number of splits is won't be exactly what we specify.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java:
```diff
@@ -0,0 +1,71 @@
+import org.joda.time.Instant;
+
+/** Unbounded poller to listen for new partitions. */
+public class PartitionPollerFn extends PollFn<Read, Read> {
```
The output of a PollFn should be something that can be efficiently and deterministically encoded (inefficiency will limit scalability; non-deterministic encoding will make Watch produce wrong results). I think in your case it could simply be an Integer for the partition index?

jhalaria commented:
@jkff - Thank you very much for your feedback. I have addressed all suggestions. Please let me know if this looks good now.

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionReaderFn.java:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  @ProcessElement
+  @SuppressWarnings("unused")
```
remove `@SuppressWarnings("unused")` everywhere

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -163,6 +176,12 @@ private HCatalogIO() {}
 
+      abstract Builder setPollingInterval(Duration pollingInterval);
+
+      abstract Builder setPartitionCols(ImmutableList<String> partitionCols);
```
Please make this a `List` to not have guava dependencies in the public API

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -186,6 +205,28 @@ public Read withFilter(String filter) {
+    }
+
+    /** Set the names of the columns that are partitions. */
+    public Read withPartitionCols(ImmutableList<String> partitionCols) {
```
same as above for List

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java:
```diff
@@ -0,0 +1,56 @@
+import org.joda.time.Instant;
+
+/** Return the list of current partitions present. */
+public class PartitionPollerFn extends PollFn<Read, Integer> {
```
Make package private

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionReaderFn.java:
```diff
@@ -0,0 +1,115 @@
+
+    final int desiredSplitCount = HCatalogUtils.getSplitCount(readRequest, partition);
+    final List<String> values = partition.getValues();
+    final ImmutableList<String> partitionCols = readRequest.getPartitionCols();
```
What happens if the user does not define partitionCols as it was before. Maybe worth to make it an empty list as default when building the Read.

jhalaria commented:
@iemejia - Feedback addressed.

iemejia commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -196,11 +237,27 @@ Read withContext(ReaderContext context) {
+      if (getPollingInterval() != null) {
+        growthFn = Watch.growthOf(new PartitionPollerFn()).withPollInterval(getPollingInterval());
+        if (getTerminationCondition() != null) {
+          growthFn =
```
Sorry I was maybe not clear, I am talking about not repeating agan the code `Watch.growthOf(new PartitionPollerFn()).withPollInterval(getPollingInterval())` because that's already done just before in the definition of `growthFn` and instead do: `growthFn = growthFn.withTerminationPerInput(getTerminationCondition());`.

It is pretty minor but nice to fix.

jhalaria replied:
Ah. I see. Done

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/HCatalogIO.java:
```diff
@@ -196,11 +253,24 @@ Read withContext(ReaderContext context) {
+            .apply("ConvertToReadRequest", Create.of(this))
+            .apply("WatchForNewPartitions", growthFn)
+            .apply("PartitionReader", ParDo.of(new PartitionReaderFn(getConfigProperties())));
+      } else {
```
Maybe add a check that getTerminationCondition() == null.

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionPollerFn.java:
```diff
@@ -0,0 +1,56 @@
+    return IntStream.range(
+            0,
+            metaStoreClient
+                .listPartitions(read.getDatabase(), read.getTable(), Short.MAX_VALUE)
```
What does Short.MAX_VALUE stand for here, and is it in any way related to Integer.MAX_VALUE above?

jhalaria replied:
listPartitions requires maximumPartitions that can be listed while making that API call. https://hive.apache.org/javadocs/r2.2.0/api/org/apache/hadoop/hive/metastore/HiveMetaStoreClient.html#listPartitions-java.lang.String-java.lang.String-short-

jkff commented on sdks/java/io/hcatalog/src/main/java/org/apache/beam/sdk/io/hcatalog/PartitionReaderFn.java:
```diff
@@ -0,0 +1,111 @@
+    final Read readRequest = c.element().getKey();
+    final Integer partitionIndexToRead = c.element().getValue();
+    ReaderContext readerContext = getReaderContext(readRequest, partitionIndexToRead);
+    for (int i = 0; i < readerContext.numSplits(); i++) {
```
Why do we need a nested loop here: is it possible to just read the whole partition without "splitting" it?

If you want to read the splits of a partition in parallel, then you need to split this into two DoFn's with a reshuffle in between.

jhalaria replied:
- As far as I know Data Transfer API(s) do not provide a way to read the whole partition.
- As far as splitting the reads into multiple DoFn's is concerned, I tried doing that but ran into an  issue which needs a fix with hive. [described below]
- Here is what I tried:

1. Pass in <ReaderContext, SplitToRead> as an output of the first DoFn.
2. Create a 2nd DoFn which reads actual data given a context and a split.
3. Added a reshuffle between the 2 DoFn(s)

This results in a `NullPointer Exception` because HCatTableInfo is transient in the ReaderContext and we don't have a way to deserialize it. I will create a PR with hcatalog to fix that. HCatTableInfo doesn't have to be transient since it is serializable.

iemejia replied:
Can you please fill the JIRA for this. Thanks!

jhalaria commented:
@jkff - Addressed feedback.

jkff commented:
LGTM, I'll let @iemejia merge though, since I'm not currently actively involved in Beam.

jhalaria commented:
@jkff @iemejia - Thank you very much for all your suggestions in making this PR a lot cleaner.

iemejia commented:
Merged manually to squash the commits.

iemejia commented:
Thanks a lot @jhalaria not only for the interesting use case but for being so eager to address the suggestions.
Thanks also @jkff it is really a pleasure to see you around giving useful feedback as before even if you are not active in the project anymore.
